src/twisted/web/http.py

```
@@ -1825,75 +1823,97 @@ class _ChunkedTransferDecoder:
+            C{self._buffer}.  C{False} when more data is required.
+
+        @raises _MalformedChunkedDataError: when the chunk size cannot be
+            decoded.
+        """
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+        if eolIndex == -1:
+            self._start = len(self._buffer) - 1
+            return False
+
+        endOfLengthIndex = self._buffer.find(b";", 0, eolIndex)
+        if endOfLengthIndex == -1:
+            endOfLengthIndex = eolIndex
+        try:
+            length = int(self._buffer[0:endOfLengthIndex], 16)
+        except ValueError:
+            raise _MalformedChunkedDataError("Chunk-size must be an integer.")
+
+        if length < 0:
+            raise _MalformedChunkedDataError("Chunk-size must not be negative.")
+        elif length == 0:
+            self.state = "TRAILER"
         else:
-            self._buffer = data
-            return b""
+            self.state = "BODY"
+
+        self.length = length
+        del self._buffer[0 : eolIndex + 2]
+        self._start = 0
+        return True
 
-    def _dataReceived_CRLF(self, data):
-        if data.startswith(b"\r\n"):
+    def _dataReceived_CRLF(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
             self.state = "CHUNK_LENGTH"
-            return data[2:]
+            del self._buffer[0:2]
+            return True
         else:
-            self._buffer = data
-            return b""
+            return False
 
-    def _dataReceived_TRAILER(self, data):
-        if data.startswith(b"\r\n"):
-            data = data[2:]
+    def _dataReceived_TRAILER(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
+            data = memoryview(self._buffer)[2:].tobytes()
+            del self._buffer[:]
             self.state = "FINISHED"
             self.finishCallback(data)
-        else:
-            self._buffer = data
-        return b""
+        return False
```

Maybe raise an exception if we are in the trailer, and the buffer and 2 or more bytes...or buffer has one byte and that byte is not `\r`

I guess that in this case the handling will raise a timeout, as the END marker is never receive.

But maybe raising an exception will prevent receiving more invalid data from the connection.